### ISSUE TITLE
Wrong MTLTextureDescriptor for texel buffer-backing textures

### ISSUE THREAD

**Issue by ch45er** [KhronosGroup/MoltenVK#542]
Hi!

When working with texel buffers in MVK, I observe the following failed Metal validation assertions:
```
-[MTLDebugBuffer newTextureWithDescriptor:offset:bytesPerRow:]:326: failed assertion `resourceOptions (0x10) must match backing buffer resource options (0x20).'
```
```
validateComputeFunctionArguments:892: failed assertion `Compute Function(Main): writes texture (...) whose usage (0x01) doesn't specify MTLTextureUsageShaderWrite (0x02)'
```

The former one occurs upon a call to `vkUpdateDescriptorSets()`, and the latter one relates to `vkQueueSubmit()`.

It seems that `id<MTLTexture> MVKBufferView::getMTLTexture()` function does not set `resourceOptions` and `usage` values in `MTLTextureDescriptor` correctly, when creates a texture:

```c++
id<MTLTexture> MVKBufferView::getMTLTexture() {
// [code ommitted for brevity]
        MTLTextureDescriptor* mtlTexDesc = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat: _mtlPixelFormat
                                                                                              width: _textureSize.width
                                                                                             height: _textureSize.height
                                                                                          mipmapped: NO];
        //mtlTexDesc.resourceOptions = 0x20;
        //mtlTexDesc.usage = MTLTextureUsageShaderWrite | MTLTextureUsageShaderRead | MTLTextureUsagePixelFormatView;
        
		_mtlTexture = [_buffer->getMTLBuffer() newTextureWithDescriptor: mtlTexDesc
																 offset: _mtlBufferOffset
															bytesPerRow: _mtlBytesPerRow];
// [code ommitted for brevity]
}
```

The commented-out lines in a middle "fix" those assertions for my particular use case.

So it seems that the `resourceOptions` and `usage` members of `mtlTexDesc` should also be set up.
 
My MVK version is `v1.0.28-280-ge771f79`, but it seems that this was present in the code since initial public release.

**Comment by billhollings** [comment]
@ch45er Thanks for the research.

I'll look into fixing this properly.

**Comment by billhollings** [comment]
@ch45er 

Can you clarify the use of your options, please? Did you actually need to set them all to get your use case satisfied?

According to Apple docs, setting `resourceOptions` shouldn't have any effect:

>If you are creating a texture whose data comes from another MTLResource object, this property value is ignored, and the value of the original resource is used instead.

And for what reason did you need `usage` to include `MTLTextureUsagePixelFormatView`? That is used for creating texture views...implying that you are intending to create a `VkImageView` from a `VkBufferView`.

**Comment by cdavis5e** [comment]
> According to Apple docs, setting `resourceOptions` shouldn't have any effect:
> 
> > If you are creating a texture whose data comes from another `MTLResource` object, this property value is ignored, and the value of the original resource is used instead.

The error message @ch45er gave implies he is running against Metal's validation layer. I wonder if it's being too strict.

**Comment by billhollings** [comment]
>I wonder if it's being too strict.

Hmmm...yes...good point. Your fix in #545 address it either way.

**Comment by ch45er** [comment]
Wow, that's alot of activity!

@billhollings 
> Can you clarify the use of your options, please? Did you actually need to set them all to get your use case satisfied?

Here's what I do:
0. create a buffer
1. pass it into a compute shader as a storage buffer
2. when the compute shader fills the buffer, use the data from it for rendering, feeding it as input to a vertex shader

I use a `VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER` descriptor for the buffer on step 1, and a `VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER` on step 2.

> According to Apple docs, setting resourceOptions shouldn't have any effect:

Yep, I've interpreted the doc the same way too, but Metal API Validation seems to say the opposite (`resourceOptions must match backing buffer resource options`). The code *seems* to work correctly if run without validation and `resourceOptions` unset. So I mostly explicitly set `resourceOptions` to make the validation happy.

> And for what reason did you need usage to include MTLTextureUsagePixelFormatView?

Oops, my bad. `MTLTextureUsagePixelFormatView` is not required in my case, must be a leftover i've overlooked. Sorry for misleading. I only create `VkBufferView`s, so this is enough:
```c++
mtlTexDesc.usage = MTLTextureUsageShaderWrite | MTLTextureUsageShaderRead;
```

Still, If I don't set `usage = MTLTextureUsageShaderWrite` explicitly, Metal validation gives me `Compute Function writes texture (...) whose usage doesn't specify MTLTextureUsageShaderWrite` error.

As @cdavis5e pointed out, these all are Metal validation errors, that appear when debugging with XCode with "Metal API Validation" enabled.
